### PR TITLE
Previews: crawler as a background service

### PR DESCRIPTION
**Pull request by ArturWierzbicki**
**What this PR does / why we need it**:

This PR sets up a scheduler so that the dashboard crawler automatically every X hours. The mechanism works on both HA and non-HA environments thanks to the [ServerLockService](https://github.com/grafana/grafana/blob/f4f0d74838ace0e9dbf2d163e031d35c96070987/pkg/infra/serverlock/serverlock.go#L18-L19)

**Which issue(s) this PR fixes**:

part of https://github.com/grafana/grafana/issues/44449

**Special notes for your reviewer**:

Functionality fixes:
- SQL query to find dashboards with stale thumbnails now takes into the account `theme` and `kind`
- fix for `light` theme crawls - dashboard URL now includes the `theme` query param
- crawler main thread now waits for all walkers to finish before broadcasting the final status update

Code fixes:
- each `pkg/services/thumbs/*.go` file has its own logger

### CODE REVIEW

File: pkg/services/thumbs/service.go [hunk 1]
```diff
@@ -284,3 +319,57 @@ func (hs *thumbService) getDashboardId(c *models.ReqContext, uid string) (int64,
...
+	}
+}
+
+func (hs *ThumbService) runScheduledCrawl(parentCtx context.Context) {
+	crawlerCtx, cancel := context.WithTimeout(parentCtx, hs.scheduleOptions.maxCrawlDuration)
+	defer cancel()
+
+	err := hs.lockService.LockAndExecute(crawlerCtx, hs.crawlLockServiceActionName, hs.scheduleOptions.crawlInterval, func(ctx context.Context) {
+		hs.renderer.Run(crawlerCtx, rendering.AuthOpts{
+			OrgID:   0,
```
Comment (ArturWierzbicki) [inline pkg/services/thumbs/service.go:344]: `{ OrgId: 0, UserID: 0}` works when I run it locally, but I am sure there is a better solution

File: pkg/services/thumbs/service.go [hunk 2]
```diff
@@ -230,22 +257,30 @@ func (hs *thumbService) StartCrawler(c *models.ReqContext) response.Response {
...
 	}
-	msg, err := hs.renderer.Start(c, cmd.Mode, cmd.Theme, models.ThumbnailKindDefault)
+
+	go hs.runOnDemandCrawl(context.Background(), cmd.Theme, cmd.Mode, models.ThumbnailKindDefault, rendering.AuthOpts{
+		OrgID:   c.OrgId,
+		UserID:  c.UserId,
+		OrgRole: c.OrgRole,
+	})
+
+	status, err := hs.renderer.Status()
```
Comment (ArturWierzbicki) [inline pkg/services/thumbs/service.go:267]: `Status()` call doesn't make sense in HA environment, I think this function can just return empty `201` and make client wait for updates from grafana live

File: pkg/services/thumbs/crawler.go [hunk 3]
```diff
@@ -62,72 +64,56 @@ func (r *simpleCrawler) next() *models.DashboardWithStaleThumbnail {
...
-	if r.status.State == running {
-		tlog.Info("already running")
-		return r.Status()
-	}
-
+func (r *simpleCrawler) Run(ctx context.Context, authOpts rendering.AuthOpts, mode CrawlerMode, theme models.Theme, thumbnailKind models.ThumbnailKind) error {
 	r.queueMutex.Lock()
-	defer r.queueMutex.Unlock()
+	if r.IsRunning() {
+		tlog.Info("Already running")
```
[the remark this answers is not in the document]
Comment (ArturWierzbicki) [reply pkg/services/thumbs/crawler.go]: fixed

File: pkg/services/thumbs/service.go [hunk 4]
```diff
@@ -284,3 +319,59 @@ func (hs *thumbService) getDashboardId(c *models.ReqContext, uid string) (int64,
...
+}
+
+func (hs *thumbService) runScheduledCrawl(parentCtx context.Context) {
+	crawlerCtx, cancel := context.WithTimeout(parentCtx, hs.scheduleOptions.maxCrawlDuration)
+	defer cancel()
+
+	authOpts := rendering.AuthOpts{
+		OrgID:   0,
+		UserID:  0,
+		OrgRole: models.ROLE_ADMIN,
```
Comment (ArturWierzbicki) [inline pkg/services/thumbs/service.go:347]: TODO: lets try with `VIEWER`

Comment (ArturWierzbicki) [reply pkg/services/thumbs/service.go]: tested - dashboards are unaccessible to _editor_ and _viewer_ org roles with `id & orgId === 0`

File: pkg/services/thumbs/crawler.go [hunk 5]
```diff
@@ -216,7 +220,7 @@ func (r *simpleCrawler) shouldWalk() bool {
 	return r.status.State == running
 }
 
-func (r *simpleCrawler) walk(ctx context.Context) {
+func (r *simpleCrawler) walk(id int, ctx context.Context) {
```
Comment (FZambia) [inline pkg/services/thumbs/crawler.go:223]: context should be first argument (just a convention in Go)

**Review by FZambia — APPROVED** [review]
👍

File: pkg/services/thumbs/crawler.go (hunk 5 above)
Comment (ArturWierzbicki) [reply pkg/services/thumbs/crawler.go]: fixed! :nail_care:

**Review by ryantxu — APPROVED** [review]
LGTM -- lots of good cleanup and has explicit enable option.

We should also move forward with setting the interval in config -- but I think we want to see how this works in the wild first :)

**Comment by ArturWierzbicki** [comment]
🚀 🚀

**Pull request merged**